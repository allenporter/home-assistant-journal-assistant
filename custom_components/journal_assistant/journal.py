"""Converter from yaml journal files to an RFC5545 Journal."""

from pathlib import Path
import datetime
import logging

from ical.calendar import Calendar
from ical.journal import Journal
from .model import JournalPage


_LOGGER = logging.getLogger(__name__)


def journal_pages(storage_dir: Path, journal_name: str) -> list[JournalPage]:
    """Load all journal pages from a storage directory with the specified journal prefix."""
    pages = []
    files = list(storage_dir.glob(f"{journal_name}-*.yaml"))
    files.sort()
    for filename in files:
        with filename.open() as file:
            page = JournalPage.from_yaml(file.read())
        pages.append(page)
    return pages


def get_dated_content(page: JournalPage) -> dict[str, list[str]]:
    """Get the date and content from a journal page."""
    default_date = page.date or page.created_at
    if not page.records:
        return {default_date: [str(page.content)]}

    dated_content: dict[str, list[str]] = {}
    for note in page.records:
        note_date = note.date or default_date
        if note_date not in dated_content:
            dated_content[note_date] = []
        dated_content[note_date].append(f"- {note.content}")
    return dated_content


def journal_from_yaml(storage_dir: Path) -> dict[str, Calendar]:
    """Convert a yaml journal to an RFC5545 Journal."""
    _LOGGER.debug("Loading journal content from %s", storage_dir)
    # Get the unique list of journal names
    journal_names = {
        filename.name.split("-")[0] for filename in storage_dir.glob("*.yaml")
    }
    _LOGGER.debug("Journal names: %s", journal_names)

    journals = {}

    keys = list(journal_names)
    keys.sort()
    for journal_name in keys:

        # Load all pages from with the same journal prefix
        pages = journal_pages(storage_dir, journal_name)

        dated_content: dict[str, list[str]] = {}
        for page in pages:
            dated_content.update(get_dated_content(page))

        if journal_name not in journals:
            journals[journal_name] = Calendar()

        # Add a journal entry for each date
        calendar = journals[journal_name]
        for date, content_list in dated_content.items():
            journal = Journal()
            journal.summary = f"{journal_name} {date}"
            if "T" in date:
                journal.dtstart = datetime.datetime.fromisoformat(date)
            else:
                journal.dtstart = datetime.date.fromisoformat(date)
            journal.description = "\n".join(content_list)
            calendar.journal.append(journal)

        journals[journal_name] = calendar

    return journals
