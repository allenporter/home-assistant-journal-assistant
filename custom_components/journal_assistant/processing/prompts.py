"""Module for loading and managing prompts."""

import logging
from functools import cache
import re
import datetime
from pathlib import Path

from .model import DynamicPrompt

_LOGGER = logging.getLogger(__name__)

DYNAMIC_PROMPTS_DIR = Path(__file__).parent / "dynamic_prompts"
DEFAULT = [
    "default",
    "rapid_log_legend",
    "profile",
]
FILE_PREFIX_PROMPT_MAP = {
    "Daily": [
        *DEFAULT,
        "daily",
    ],
    "Weekly": [
        *DEFAULT,
        "weekly",
    ],
    "Monthly": [
        *DEFAULT,
        "monthly",
    ],
}

TIMESTAMP_RE = re.compile(r".*?-\d+-P(\d{20}).*")

FILE_PROMPT = """Please answer in json with no other formatting since the answer will be parsed programmatically.

Filename: {filename}
Created At: {created_at}
Content:
"""


@cache
def _load_dynamic_prompts() -> dict[str, DynamicPrompt]:
    """Load all dynamic prompts from the prompts directory."""

    dynamic_prompts = {}
    for filename in DYNAMIC_PROMPTS_DIR.glob("*.yaml"):
        _LOGGER.info(f"Loading dynamic prompt: {filename}")
        dynamic_prompts[filename.stem] = DynamicPrompt.from_file(filename)
    return dynamic_prompts


def get_dynamic_prompts(page_filename: Path) -> list[DynamicPrompt]:
    """Get a set of prompts that match the given prefix"""

    page_name = page_filename.stem
    prefix = page_name.split("-")[0]
    prompt_names = FILE_PREFIX_PROMPT_MAP.get(prefix, DEFAULT)
    dynamic_prompts = _load_dynamic_prompts()
    return [
        value
        for prompt_prefix in prompt_names
        for key, value in dynamic_prompts.items()
        if key.startswith(prompt_prefix)
    ]


def get_file_prompt(path: Path) -> str:
    """Get a prompt for a file."""
    re_match = TIMESTAMP_RE.match(str(path.name))
    if re_match is not None:
        created_at = datetime.datetime.strptime(re_match.group(1), "%Y%m%d%H%M%S%f")
    else:
        created_at = datetime.datetime.now()
    return FILE_PROMPT.format(
        filename=str(path.name), created_at=created_at.isoformat()
    )
