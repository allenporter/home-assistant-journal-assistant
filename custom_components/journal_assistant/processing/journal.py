"""Converter from yaml journal files to an RFC5545 Journal."""

from pathlib import Path
import datetime
import logging
import hashlib
from typing import cast

from ical.calendar import Calendar
from ical.journal import Journal

from .model import JournalPage

_LOGGER = logging.getLogger(__name__)

__all__ = [
    "journal_from_yaml",
]


def journal_pages(storage_dir: Path, journal_name: str) -> list[JournalPage]:
    """Load all journal pages from a storage directory with the specified journal prefix."""
    pages = []
    files = list(storage_dir.glob(f"{journal_name}-*.yaml"))
    files.sort()
    for filename in files:
        with filename.open() as file:
            page = JournalPage.from_yaml(file.read())
        pages.append(page)
    return pages


def get_dated_content(page: JournalPage) -> dict[str, list[str]]:
    """Get the date and content from a journal page."""
    default_date = page.date or page.created_at
    if not page.records:
        return {default_date: [str(page.content)]}

    dated_content: dict[str, list[str]] = {}
    for note in page.records:
        note_date = note.date or default_date
        if note_date not in dated_content:
            dated_content[note_date] = []
        content = ""
        if note.content:
            if note.status:
                content = f"- ({note.status}) {note.content}"
            else:
                content = f"- {note.content}"
        if note.entries:
            if content:
                content += "\n"
            content += "\n".join(f"  - {entry}" for entry in note.entries if entry)
        dated_content[note_date].append(content)
    return dated_content


def write_content(content: str, filename: Path) -> None:
    """Write content to a file."""
    with filename.open("w") as file:
        file.write(content)


def write_journal_page_yaml(
    storage_dir: Path,
    note_name: str,
    page: JournalPage,
) -> None:
    """Write a journal page yaml string to a yaml file."""
    if "-" not in note_name:
        raise ValueError(f"Note name must contain a '-' character: {note_name}")
    if not page.filename.startswith(note_name):
        raise ValueError(
            f"Note name must match page filename: {note_name} != {page.filename}"
        )
    storage_dir.mkdir(parents=True, exist_ok=True)
    filename = storage_dir / f"{note_name}.yaml"
    _LOGGER.debug("Writing journal page to %s", filename)
    content = cast(str, page.to_yaml())
    write_content(content, filename)


def journal_from_yaml(
    storage_dir: Path,
    allowed_notes: set[str],
    default_note_name: str,
) -> dict[str, Calendar]:
    """Convert a yaml journal to an RFC5545 Journal."""
    _LOGGER.debug("Loading journal content from %s", storage_dir)
    # Get the unique list of journal names
    filename_prefixes = {
        filename.name.split("-")[0] for filename in storage_dir.glob("*.yaml")
    }
    note_names = list(filename_prefixes)
    note_names.sort()
    _LOGGER.debug("Journal names: %s", note_names)

    journals = {}
    for note_name in note_names:

        # Load all pages from with the same journal prefix
        pages = journal_pages(storage_dir, note_name)

        # Allow notes to have their own calendar entry if in the list of allowed notes
        key_name = note_name if note_name in allowed_notes else default_note_name

        dated_content: dict[str, list[str]] = {}
        for page in pages:
            for date, content_list in get_dated_content(page).items():
                if date not in dated_content:
                    dated_content[date] = []
                dated_content[date].extend(content_list)

        if key_name not in journals:
            journals[key_name] = Calendar()

        # Add a journal entry for each date
        calendar = journals[key_name]
        for date, content_list in dated_content.items():
            journal = Journal()
            journal.uid = hashlib.sha256(f"{note_name}-{date}".encode()).hexdigest()
            journal.summary = f"{note_name} {date}"
            journal.categories = [key_name]
            if note_name not in allowed_notes:
                journal.categories.append(note_name)
            if "T" in date:
                journal.dtstart = datetime.datetime.fromisoformat(date)
            else:
                journal.dtstart = datetime.date.fromisoformat(date)
            journal.description = "\n".join(content_list)
            calendar.journal.append(journal)

        journals[key_name] = calendar

    return journals
