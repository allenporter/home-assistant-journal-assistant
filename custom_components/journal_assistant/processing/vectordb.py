"""Journal Assistant vector search database."""

import itertools
import logging
from typing import Any
from pathlib import Path
import yaml


import chromadb
from chromadb.api.types import IncludeEnum
from chromadb.config import DEFAULT_TENANT, DEFAULT_DATABASE, Settings
from chromadb.utils.embedding_functions import google_embedding_function

from ical.calendar import Calendar
from ical.journal import Journal


_LOGGER = logging.getLogger(__name__)

BATCH_SIZE = 10
COLLECTION_NAME = "journal_assistant"
MODEL = "models/text-embedding-004"


def serialize_content(item: Journal) -> str:
    """Serialize a journal entry."""
    return yaml.dump(item.dict(exclude={"uid", "dtsamp"}, exclude_unset=True, exclude_none=True))  # type: ignore[no-any-return]


class VectorDB:
    """Journal Assistant vector search database."""

    def __init__(self, storage_path: Path, google_api_key: str) -> None:
        """Initialize the vector database."""
        _LOGGER.debug("Creating ChromaDB System")
        settings = Settings(
            anonymized_telemetry=False,
        )
        self.system = chromadb.config.System(settings)
        _LOGGER.debug("Creating Google embedding function")
        # Separate embedding functions are used for idnexing vs querying
        self.index_embedding_function = (
            google_embedding_function.GoogleGenerativeAiEmbeddingFunction(
                api_key=google_api_key, model_name=MODEL, task_type="RETRIEVAL_DOCUMENT"
            )
        )
        self.query_embedding_function = (
            google_embedding_function.GoogleGenerativeAiEmbeddingFunction(
                api_key=google_api_key, model_name=MODEL, task_type="RETRIEVAL_QUERY"
            )
        )
        self.client = chromadb.PersistentClient(
            path=str(storage_path),
            settings=settings,
            tenant=DEFAULT_TENANT,
            database=DEFAULT_DATABASE,
        )
        _LOGGER.debug("Creating collection: %s", COLLECTION_NAME)

    def upsert_index(self, notebooks: dict[str, Calendar]) -> None:
        """Add notebooks to the index."""
        _LOGGER.debug("Adding notebooks to index")

        collection = self.client.get_or_create_collection(
            name=COLLECTION_NAME, embedding_function=self.index_embedding_function  # type: ignore[arg-type]
        )

        for note_name, calendar in notebooks.items():
            _LOGGER.debug(
                "Indexing %s with %s entries", note_name, len(calendar.journal)
            )
            for found_journal_entries in itertools.batched(
                calendar.journal, BATCH_SIZE
            ):
                _LOGGER.debug(
                    "Examining batch of %s to index", len(found_journal_entries)
                )
                ids = [
                    journal_entry.uid or "" for journal_entry in found_journal_entries
                ]
                results = collection.get(ids=ids, include=[IncludeEnum.documents])
                _LOGGER.debug("Results: %s", results)
                existing_ids = {
                    uid: documents
                    for uid, documents in zip(
                        results["ids"], results.get("documents") or []
                    )
                }
                if existing_ids:
                    _LOGGER.debug("Found %s existing documents", len(existing_ids))
                # Determine which journal entries are entirely new or have
                # updated descriptions
                journal_entries = []
                documents = []
                metadatas = []
                for journal_entry in found_journal_entries:
                    existing_content = existing_ids.get(journal_entry.uid)
                    entry_content = serialize_content(journal_entry)
                    if existing_content is None or entry_content != existing_content:
                        journal_entries.append(journal_entry)
                        documents.append(entry_content)
                        metadatas.append(
                            {
                                "category": (
                                    journal_entry.categories[0]
                                    if journal_entry.categories
                                    else ""
                                ),
                                "date": journal_entry.dtstart.isoformat(),
                                "name": journal_entry.summary or "",
                            }
                        )
                ids = [journal_entry.uid or "" for journal_entry in journal_entries]
                if not ids:
                    _LOGGER.debug("Skipping batch of unchanged documents")
                    continue
                _LOGGER.debug("Upserting batch of %s to index", len(journal_entries))
                collection.upsert(
                    documents=documents,
                    metadatas=metadatas,  # type: ignore[arg-type]
                    ids=ids,
                )

    def count(self) -> int:
        """Return the number of documents in the collection."""
        collection = self.client.get_collection(
            name=COLLECTION_NAME, embedding_function=self.query_embedding_function  # type: ignore[arg-type]
        )
        return collection.count()

    def query(self, query: str, num_results: int) -> list[dict[str, Any]]:
        """Search the VectorDB for relevant documents."""
        collection = self.client.get_collection(
            name=COLLECTION_NAME, embedding_function=self.query_embedding_function  # type: ignore[arg-type]
        )

        results: chromadb.QueryResult = collection.query(
            query_texts=query,
            n_results=num_results,
        )
        if (
            (id_list := results.get("ids")) is None
            or (metadata_list := results.get("metadatas")) is None
            or (document_list := results.get("documents")) is None
        ):
            raise ValueError(f"Invalid query results: {results}")
        # Get the results for the first (and only) query
        ids = id_list[0]
        metadatas = metadata_list[0]
        documents = document_list[0]
        return [
            {
                "id": ids[index],
                "content": documents[index],
                **metadatas[index],
            }
            for index in range(len(ids))
        ]
