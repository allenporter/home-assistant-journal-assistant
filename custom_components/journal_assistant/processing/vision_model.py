"""Multi-modal vision model for processing journal pages."""

import io
import re
import logging
import json
import yaml
from typing import cast
import datetime
from pathlib import Path

import google.generativeai as genai
import PIL.Image
from mashumaro.exceptions import MissingField

from .prompts import get_dynamic_prompts
from .model import JournalPage

_LOGGER = logging.getLogger(__name__)

TIMESTAMP_RE = re.compile(r".*?-\d+-P(\d{20}).*.png")

FILE_PROMPT = """
Please answer in json with no other formatting since the answer will be parsed programmatically.

Filename: {filename}
Created At: {created_at}
Content:
"""
EXTRACT_JSON = re.compile("```json\n(.*?)\n```", re.DOTALL)


def _parse_model_response(response_text: str) -> str:
    """Parse the response from the model and return a yaml string."""

    if (match := EXTRACT_JSON.match(response_text)) is not None:
        text = match.group(1)
    else:
        text = response_text

    try:
        obj = json.loads(text)
    except ValueError as err:
        _LOGGER.error("Error processing: %s", err)
        return text

    for k, v in list(obj.items()):
        if v is None or v == "null":
            del obj[k]

    return cast(str, yaml.dump(obj, explicit_start=True, sort_keys=False))


class VisionModel:
    """Multi-modal vision model for processing journal pages."""

    def __init__(self, model: genai.GenerativeModel) -> None:
        """Initialize the vision model."""
        self._model = model

    async def process_journal_page(
        self, page_name: Path, page_content: bytes
    ) -> JournalPage:
        """Process a journal page using a multi-modal vision model.

        The response is typically a yaml content string.
        """
        _LOGGER.debug("Extract content from page %s", str(page_name))

        created_at: datetime.datetime | None = None
        if (re_match := TIMESTAMP_RE.match(str(page_name))) is not None:
            created_at = datetime.datetime.strptime(re_match.group(1), "%Y%m%d%H%M%S%f")

        prompts = get_dynamic_prompts(page_name)
        prompt = "\n\n".join([p.as_prompt() for p in prompts])

        img = PIL.Image.open(io.BytesIO(page_content))

        response = await self._model.generate_content_async(
            [
                prompt,
                FILE_PROMPT.format(
                    filename=f"{page_name.stem}.png",
                    created_at=created_at.isoformat() if created_at else "N/A",
                ),
                img,
            ]
        )

        yaml_content = _parse_model_response(response.text)
        try:
            return JournalPage.from_yaml(yaml_content)
        except MissingField as err:
            raise ValueError(f"Error parsing journal page: {err}")
