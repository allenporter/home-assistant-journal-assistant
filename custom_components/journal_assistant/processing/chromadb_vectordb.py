"""Journal Assistant vector search database."""

import logging
from typing import Any
from pathlib import Path
from urllib.parse import urlparse
import datetime

import chromadb
from chromadb.errors import ChromaError
from chromadb.api.types import IncludeEnum
from chromadb.config import DEFAULT_TENANT, DEFAULT_DATABASE, Settings
from chromadb.utils.embedding_functions import google_embedding_function

from homeassistant.core import HomeAssistant

from custom_components.journal_assistant.vectordb import (
    VectorDB,
    IndexableDocument,
    QueryParams,
    VectorDBError,
    QueryResult,
)

_LOGGER = logging.getLogger(__name__)


DEFAULT_MAX_RESULTS = 10
COLLECTION_NAME = "journal_assistant"
MODEL = "models/text-embedding-004"
EMPTY_QUERY = "task"  # Arbitrary query to use when no query is provided


def _as_query_args(params: QueryParams) -> dict[str, Any]:
    """Return the query arguments for vectordb query."""
    args: dict[str, Any] = {}
    args["query_texts"] = [params.query if params.query else EMPTY_QUERY]
    filters: list[dict[str, Any]] = []
    if params.metadata is not None:
        filters.append(params.metadata)
    if params.start_date is not None:
        filters.append({"date": {"$gte": params.start_date.timestamp()}})
    if params.end_date is not None:
        filters.append({"date": {"$lte": params.end_date.timestamp()}})
    if len(filters) > 1:
        args["where"] = {"$and": filters}
    elif len(filters) == 1:
        args["where"] = filters[0]
    args["n_results"] = params.num_results or DEFAULT_MAX_RESULTS
    return args


def _create_chromadb_client(chromadb_url: str, tenant: str) -> chromadb.api.ClientAPI:
    """Create a ChromaDB client."""
    _LOGGER.debug("Creating ChromaDB client for tenant: %s", tenant)
    url = urlparse(chromadb_url)
    port = 80
    if url.port:
        port = url.port
    elif url.scheme == "https":
        port = 443
    return chromadb.HttpClient(
        settings=Settings(
            anonymized_telemetry=False,
        ),
        host=url.hostname or "",
        port=port,
        ssl=(url.scheme == "https"),
        tenant=tenant,
        database=DEFAULT_DATABASE,
    )


def create_tenant(chromadb_url: str, tenant: str) -> None:
    """Get or create a tenant."""
    url = urlparse(chromadb_url)
    port = 80
    if url.port:
        port = url.port
    elif url.scheme == "https":
        port = 443
    settings = Settings(
        chroma_api_impl="chromadb.api.fastapi.FastAPI",
        chroma_server_host=url.hostname,
        chroma_server_http_port=port,
        chroma_server_ssl_enabled=(url.scheme == "https"),
        anonymized_telemetry=False,
    )
    _LOGGER.debug("Creating tenant: %s", tenant)
    admin_client = chromadb.AdminClient(settings=settings)
    admin_client.create_tenant(tenant)
    admin_client.create_database(DEFAULT_DATABASE, tenant)
    _LOGGER.debug("Tenant created: %s", tenant)


def create_local_chroma_client(storage_path: Path) -> chromadb.api.ClientAPI:
    """Create a ChromaDB client."""
    return chromadb.PersistentClient(
        path=str(storage_path),
        settings=Settings(
            anonymized_telemetry=False,
        ),
        tenant=DEFAULT_TENANT,
        database=DEFAULT_DATABASE,
    )


class ChromaVectorDB:
    """Journal Assistant vector search database."""

    def __init__(self, client: chromadb.api.ClientAPI, google_api_key: str) -> None:
        """Initialize the vector database."""
        _LOGGER.debug("Creating ChromaDB System")
        self.client = client
        _LOGGER.debug("Creating Google embedding function")
        # Separate embedding functions are used for idnexing vs querying
        self.index_embedding_function = (
            google_embedding_function.GoogleGenerativeAiEmbeddingFunction(
                api_key=google_api_key, model_name=MODEL, task_type="RETRIEVAL_DOCUMENT"
            )
        )
        self.query_embedding_function = (
            google_embedding_function.GoogleGenerativeAiEmbeddingFunction(
                api_key=google_api_key, model_name=MODEL, task_type="RETRIEVAL_QUERY"
            )
        )
        _LOGGER.debug("Creating collection: %s", COLLECTION_NAME)

    def _query_collection(self) -> chromadb.Collection:
        """Get the collection for the VectorDB."""
        return self.client.get_or_create_collection(
            name=COLLECTION_NAME, embedding_function=self.query_embedding_function  # type: ignore[arg-type]
        )

    def _index_collection(self) -> chromadb.Collection:
        """Get the collection for the VectorDB."""
        return self.client.get_or_create_collection(
            name=COLLECTION_NAME, embedding_function=self.index_embedding_function  # type: ignore[arg-type]
        )

    def upsert_index(self, documents: list[IndexableDocument]) -> None:
        """Add notebooks to the index."""
        _LOGGER.debug("Upserting %d documents in the index", len(documents))

        collection = self._index_collection()
        ids = [document.uid for document in documents]

        results = collection.get(ids=ids, include=[IncludeEnum.documents])
        existing_ids = {
            uid: documents
            for uid, documents in zip(results["ids"], results.get("documents") or [])
        }
        if existing_ids:
            _LOGGER.debug("Found %s existing documents", len(existing_ids))
        # Determine which documents entries are entirely new or have updated content
        upsert_documents = []
        for document in documents:
            if (
                existing_content := existing_ids.get(document.uid)
            ) is None or document.document != existing_content:
                upsert_documents.append(document)
        if not upsert_documents:
            _LOGGER.debug("Skipping batch of unchanged documents")
            return
        metadatas: list[dict[str, str]] = []
        for document in upsert_documents:
            metadata = {
                **document.metadata,
            }
            if document.timestamp:
                metadata["date"] = document.timestamp.timestamp()
            metadatas.append(metadata)
        collection.upsert(
            documents=[document.document for document in upsert_documents],
            metadatas=metadatas,  # type: ignore[arg-type]
            ids=[document.uid for document in upsert_documents],
        )

    def count(self) -> int:
        """Return the number of documents in the collection."""
        collection = self._query_collection()
        return collection.count()

    def query(self, params: QueryParams) -> list[QueryResult]:
        """Search the VectorDB for relevant documents."""
        collection = self._query_collection()
        query_args = _as_query_args(params)
        _LOGGER.debug("Querying collection %s with args %s", collection, query_args)
        results: chromadb.QueryResult = collection.query(**query_args)
        if (
            (id_list := results.get("ids")) is None
            or (metadata_list := results.get("metadatas")) is None
            or (document_list := results.get("documents")) is None
            or (distance_list := results.get("distances")) is None
        ):
            raise ValueError(f"Invalid query results: {results}")
        # Get the results for the first (and only) query
        ids = id_list[0]
        metadatas = metadata_list[0]
        documents = document_list[0]
        distances = distance_list[0]
        return [
            QueryResult(
                document=IndexableDocument(
                    uid=ids[index],
                    document=documents[index],
                    timestamp=datetime.datetime.fromtimestamp(
                        float(metadatas[index].get("date"))  # type: ignore[arg-type]
                    ),
                    metadata={**metadatas[index]},
                ),
                score=distances[index],
            )
            for index in range(len(ids))
        ]


class AsyncChromaVectorDB(VectorDB):
    """Journal Assistant vector search database."""

    def __init__(self, hass: HomeAssistant, chroma_vectordb: ChromaVectorDB) -> None:
        """Initialize the vector database."""
        self._hass = hass
        self._chroma_vectordb = chroma_vectordb

    async def upsert_index(self, documents: list[IndexableDocument]) -> None:
        """Add notebooks to the index."""
        await self._hass.async_add_executor_job(
            self._chroma_vectordb.upsert_index, documents
        )

    async def count(self) -> int:
        """Return the number of documents in the collection."""
        return self._chroma_vectordb.count()

    async def query(self, params: QueryParams) -> list[QueryResult]:
        """Search the VectorDB for relevant documents."""
        return await self._hass.async_add_executor_job(
            self._chroma_vectordb.query, params
        )


def create_chroma_db(
    hass: HomeAssistant,
    chromadb_url: str,
    tenant: str,
    api_key: str,
) -> VectorDB:
    _LOGGER.debug("Creating VectorDB")
    try:
        client = _create_chromadb_client(chromadb_url, tenant)
    except ChromaError as err:
        _LOGGER.error("Error creating ChromaDB client: %s", err)
        raise VectorDBError(f"Error creating ChromaDB client: {err}") from err
    db = ChromaVectorDB(client, api_key)
    return AsyncChromaVectorDB(hass, db)
