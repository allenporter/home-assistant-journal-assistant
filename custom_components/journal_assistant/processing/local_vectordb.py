"""Journal Assistant vector search database."""

import itertools
import logging
from dataclasses import dataclass
from collections.abc import Callable, Awaitable

import numpy as np


from custom_components.journal_assistant.vectordb import (
    VectorDB,
    IndexableDocument,
    QueryParams,
    QueryResult,
)

_LOGGER = logging.getLogger(__name__)


DEFAULT_MAX_RESULTS = 10
COLLECTION_NAME = "journal_assistant"
MODEL = "models/text-embedding-004"
EMPTY_QUERY = "task"  # Arbitrary query to use when no query is provided


@dataclass(kw_only=True)
class Embedding:
    """An embedding."""

    embedding: np.ndarray
    """The embedding."""


EmbeddingFunction = Callable[[str], Awaitable[Embedding]]


class LocalVectorDB(VectorDB):
    """Local vector search database."""

    def __init__(
        self, index_fn: EmbeddingFunction, query_fn: EmbeddingFunction
    ) -> None:
        """Initialize the vector database."""
        self._index_fn = index_fn
        self._query_fn = query_fn
        self._documents: dict[str, IndexableDocument] = {}
        self._embeddings: dict[str, Embedding] = {}

    async def upsert_index(self, documents: list[IndexableDocument]) -> None:
        """Add notebooks to the index."""
        _LOGGER.debug("Upserting %d documents in the index", len(documents))

        for document in documents:
            if existing_document := self._documents.get(document.uid):
                if existing_document.timestamp == document.timestamp:
                    # Skip if the document is already in the index
                    continue
            self._documents[document.uid] = document
            self._embeddings[document.uid] = await self._index_fn(document.document)

    async def count(self) -> int:
        """Return the number of documents in the collection."""
        return len(self._documents)

    async def query(self, params: QueryParams) -> list[QueryResult]:
        """Search the VectorDB for relevant documents."""

        # The results will be sorted by the query embedding
        query_embedding: Embedding | None = None
        if params.query:
            query_embedding = await self._query_fn(params.query)

        def document_filter(document: IndexableDocument) -> bool:
            if params.start_date is not None and (
                document.timestamp is None or document.timestamp < params.start_date
            ):
                return False
            if params.end_date is not None and (
                document.timestamp is None or document.timestamp > params.end_date
            ):
                return False
            if params.metadata is not None:
                for key, value in params.metadata.items():
                    if document.metadata.get(key) != value:
                        return False
            return True

        def compute_distance(document: IndexableDocument) -> float:
            if query_embedding is None:
                return 0.0
            return np.linalg.norm(  # type: ignore[return-value]
                query_embedding.embedding - self._embeddings[document.uid].embedding
            )

        distances = sorted(
            (
                QueryResult(
                    score=compute_distance(document),
                    document=document,
                )
                for document in self._documents.values()
                if document_filter(document)
            ),
            key=lambda result: result.score,
        )
        return list(
            itertools.islice(distances, params.num_results or DEFAULT_MAX_RESULTS)
        )
