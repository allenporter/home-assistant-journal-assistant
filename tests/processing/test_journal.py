"""Test parsing journal entries as a RFC5545 Journal."""

from pathlib import Path

from syrupy import SnapshotAssertion

from custom_components.journal_assistant.processing.journal import journal_from_yaml


def test_parse_journal_as_calendar(snapshot: SnapshotAssertion) -> None:
    """Test parsing a journal page."""

    calendars = journal_from_yaml(
        Path("tests/fixtures"), {"Daily", "Monthly"}, "Journal"
    )
    assert calendars.keys() == {"Daily", "Journal", "Monthly"}
    assert [
        entry.dict(exclude={"uid", "dtstamp"}, exclude_none=True, exclude_unset=True)
        for calendar in calendars.values()
        for entry in calendar.journal
    ] == snapshot
