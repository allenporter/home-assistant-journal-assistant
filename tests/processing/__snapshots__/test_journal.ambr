# serializer version: 1
# name: test_parse_journal_as_calendar
  list([
    dict({
      'categories': list([
        'Daily',
      ]),
      'description': '''
        - (completed) migrate Dec to supernote
        - set Plans w/ mom
        - (migrated) call Dad re xmas plan
        - gifts info to mom & Sam
        - prod readiness pass
        - (migrated) xmas tree
        - Dot template: darker 5mm
      ''',
      'dtstart': datetime.date(2023, 12, 19),
      'summary': 'Daily 2023-12-19',
    }),
    dict({
      'categories': list([
        'Daily',
      ]),
      'description': '''
        - (migrated) xmas tree
        - text mom for xmas day
        - call Dad: xmas plans
        - (migrated) gifts plan w/ Cathrine
        - recognition vs old notes
        - living room clutter
        - okay of the house activity
        - (migrated) super note box clean up
        - flux-local helm magic
      ''',
      'dtstart': datetime.date(2023, 12, 20),
      'summary': 'Daily 2023-12-20',
    }),
    dict({
      'categories': list([
        'Daily',
      ]),
      'description': '''
        - cardboard breakdown
        - (migrated) Bowling w/ Q
        - (completed) flux-local helm
        - todo urls?
        - (migrated) gifts plan
        - windows xmas lights
        - (migrated) fitbit python
      ''',
      'dtstart': datetime.date(2023, 12, 21),
      'summary': 'Daily 2023-12-21',
    }),
    dict({
      'categories': list([
        'Daily',
      ]),
      'description': '- clean garage',
      'dtstart': datetime.date(2023, 12, 22),
      'summary': 'Daily 2023-12-22',
    }),
    dict({
      'categories': list([
        'Journal',
        'Homelab',
      ]),
      'description': '''
        X Cluster Rebuild
        X external dns config
        X kairos image
        X adguard on campus
      ''',
      'dtstart': datetime.datetime(2023, 12, 23, 17, 18, 11, 138118),
      'summary': 'Homelab 2023-12-23T17:18:11.138118',
    }),
    dict({
      'categories': list([
        'Journal',
        'Weekly',
      ]),
      'description': '''
        - What progress did I make last week?
          - established working group
          - performance sessions complete
          - Knott's Berry farm trip
          - Finished recording AOT tracks
      ''',
      'dtstart': datetime.datetime(2024, 10, 7, 7, 13, 23, 418273),
      'summary': 'Weekly 2024-10-07T07:13:23.418273',
    }),
    dict({
      'categories': list([
        'Monthly',
      ]),
      'description': '''
        - How am I doing on my objectives and relationships?
        - Continuing to read papers and other development objectives
        - Had good quality time with family
      ''',
      'dtstart': datetime.date(2024, 4, 1),
      'summary': 'Monthly 2024-04-01',
    }),
  ])
# ---
